### PR TITLE
Adding a permission matrix

### PR DESCRIPTION
**Pull request by ThibautAlgrin**

### CODE REVIEW

File: Resources/public/js/permission.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+$().ready(function () {
```
Comment (rande) [inline Resources/public/js/permission.js]: you need to use ``jQuery`` and not ``$``

**Comment by rande** [comment]
Thanks for contributing, however before we go further, some points need to be addressed:
- You need to add licence header on each files
- You need to update tests and documentation
- The code does not respect Symfony2's CS.
- Also the code does not seems to be BC so you need to explain in the change log how to update the code. 
- An explanation and a screenshot will be welcomed in the PR's comment.

**Comment by OskarStark** [comment]
And have a look at you first command, this one is not referenced to your username ;-)

bildschirmfoto 2015-04-29 um 10 44 54

**Comment by rande** [comment]
@OskarStark I think @ThibautAlgrin just want to contribute his internal work to Sonata.

**Comment by OskarStark** [comment]
@rande could be possible, but maybe he will fix the commits to be named in the project ;-)
just as hint, not as a duty

**Comment by rande** [comment]
Yes, it will be possible to squash commits once the PR is ready.

File: Resources/doc/reference/permission_matrix.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    - Export
+
+I add two permissions:
+    - Operator: you can manage this entity
+    - Master: you can manage this entity and delegate permission
+
+
+The matrix goes represented these permissions in table, as shows us this screen shot
+
+![The premission's matrice](../images/premission_matrice.png)
```
Comment (OskarStark) [inline Resources/doc/reference/permission_matrix.rst]: should be in English, the filename, too

File: Resources/doc/reference/permission_matrix.rst [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+You have two parts:
+ - Six default permissions for each entity
+ - Customes roles defined in the ``security.yml``
+
+It's important to know, if you define a new permission in your ``security.yml``, it will be automatically add in the second part
```
Comment (OskarStark) [inline Resources/doc/reference/permission_matrix.rst]: , it will be added automatically...

File: Resources/doc/reference/permission_matrix.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+How to use it
+-------------
+
+You need to use ``sonata_security_roles`` form type.
```
Comment (OskarStark) [inline Resources/doc/reference/permission_matrix.rst]: You need to use THE ``sonata_security_roles``form type.

File: Security/EditableRolesBuilder.php [hunk 5]
```diff
@@ -17,11 +17,47 @@
...
+    /**
+     * @var SecurityContextInterface
+     */
     protected $securityContext;
-
+    /**
+     * @var Pool
+     */
     protected $pool;
-
```
Comment (OskarStark) [inline Security/EditableRolesBuilder.php]: keep the new line

File: Security/EditableRolesBuilder.php [hunk 6]
```diff
@@ -17,11 +17,47 @@
 
 class EditableRolesBuilder
 {
+    /**
+     * @var SecurityContextInterface
+     */
     protected $securityContext;
-
```
Comment (OskarStark) [inline Security/EditableRolesBuilder.php]: keep the new line

File: Form/Type/SecurityRolesType.php [hunk 7]
```diff
@@ -28,7 +28,6 @@
 class SecurityRolesType extends AbstractType
 {
     protected $rolesBuilder;
-
```
Comment (OskarStark) [inline Form/Type/SecurityRolesType.php]: keep the new line

File: Form/Type/SecurityRolesType.php [hunk 8]
```diff
@@ -36,69 +35,38 @@ public function __construct(EditableRolesBuilder $rolesBuilder)
     {
         $this->rolesBuilder = $rolesBuilder;
     }
-
```
Comment (OskarStark) [inline Form/Type/SecurityRolesType.php]: keep the new line

File: Form/Type/SecurityRolesType.php [hunk 9]
```diff
@@ -36,69 +35,38 @@ public function __construct(EditableRolesBuilder $rolesBuilder)
...
-        if (isset($attr['class']) && empty($attr['class'])) {
-            $attr['class'] = 'sonata-medium';
-        }
-
-        $view->vars['attr'] = $attr;
         $view->vars['read_only_choices'] = $options['read_only_choices'];
+        $view->vars['labelPermission'] = $this->rolesBuilder->getLabelPermission();
+        $view->vars['labelAdmin'] = $this->rolesBuilder->getLabelAdmin();
     }
-
```
Comment (OskarStark) [inline Form/Type/SecurityRolesType.php]: keep the new line

File: Form/Type/SecurityRolesType.php [hunk 10]
```diff
@@ -36,69 +35,38 @@ public function __construct(EditableRolesBuilder $rolesBuilder)
...
-            'data_class' => null
+            'read_only_choices' => $rolesReadOnly,
+            'data_class' => null,
+            'attr' => array(
+                'class' => '',
+            ),
         ));
     }
-
```
Comment (OskarStark) [inline Form/Type/SecurityRolesType.php]: keep the new line

File: Resources/config/twig.xml [hunk 11]
```diff
@@ -8,5 +8,8 @@
         <service id="sonata.user.twig.global" class="Sonata\UserBundle\Twig\GlobalVariables" >
             <argument type="service" id="service_container" />
         </service>
+        <service id="sonata.user.twig.extentions" class="Sonata\UserBundle\Twig\SecurityExtentions">
```
Comment (OskarStark) [inline Resources/config/twig.xml]: extenTion??? shoudl be extenSion

and

Sonata\UserBundle\Twig\SecurityExtentions should be Sonata\UserBundle\Twig\SecurityExtensions

File: Security/EditableRolesBuilder.php [hunk 12]
```diff
@@ -17,11 +17,47 @@
...
+    /**
+     * @var Pool
+     */
     protected $pool;
-
+    /**
+     * @var array
+     */
     protected $rolesHierarchy;
+    /**
```
Comment (OskarStark) [inline Security/EditableRolesBuilder.php]: add new lines after each property

File: Security/EditableRolesBuilder.php [hunk 13]
```diff
@@ -17,11 +17,47 @@
...
+     * @param $needle
+     * @param array $haystack
+     * @return bool
+     */
+    private function recursiveArraySearch($needle, array $haystack)
+    {
+        foreach ($haystack as $key => $value)
+        {
+            if($needle === $key || (is_array($value) && $this->recursiveArraySearch($needle,$value) === true))
```
Comment (OskarStark) [inline Security/EditableRolesBuilder.php]: please execute ``make cs``to fix the CodeStyle

File: Twig/SecurityExtentions.php [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ */
+
+namespace Sonata\UserBundle\Twig;
+
+
+class SecurityExtentions extends \Twig_Extension {
```
Comment (OskarStark) [inline Twig/SecurityExtentions.php]: Should be ``SecurityExtension`` IMO

File: Twig/SecurityExtentions.php [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return 'sizannia_jquery_tools_extension';
+    }
+
+    /**
+     * @param $value
+     * @param array $_array
+     * @return bool
+     */
+    public function isReadOnly($value, array $_array)
```
Comment (OskarStark) [inline Twig/SecurityExtentions.php]: ``$_array``-> ``$array``

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Pull request closed without merging**